File: ts/src/nodes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const syntax_1 = require("./syntax");
/* tslint:disable:max-classes-per-file */
class ArrayExpression {
    constructor(elements) {
        this.type = syntax_1.Syntax.ArrayExpression;
        this.elements = elements;
    }
}
exports.ArrayExpression = ArrayExpression;
class ArrayPattern {
    constructor(elements) {
        this.type = syntax_1.Syntax.ArrayPattern;
        this.elements = elements;
    }
}
exports.ArrayPattern = ArrayPattern;
class ArrowFunctionExpression {
    constructor(params, body, expression) {
        this.type = syntax_1.Syntax.ArrowFunctionExpression;
        this.id = null;
        this.params = params;
        this.body = body;
        this.generator = false;
        this.expression = expression;
        this.async = false;
    }
}
exports.ArrowFunctionExpression = ArrowFunctionExpression;
class AssignmentExpression {
    constructor(operator, left, right) {
        this.type = syntax_1.Syntax.AssignmentExpression;
        this.operator = operator;
        this.left = left;
        this.right = right;
    }
}
exports.AssignmentExpression = AssignmentExpression;
class AssignmentPattern {
    constructor(left, right) {
        this.type = syntax_1.Syntax.AssignmentPattern;
        this.left = left;
        this.right = right;
    }
}
exports.AssignmentPattern = AssignmentPattern;
class AsyncArrowFunctionExpression {
    constructor(params, body, expression) {
        this.type = syntax_1.Syntax.ArrowFunctionExpression;
        this.id = null;
        this.params = params;
        this.body = body;
        this.generator = false;
        this.expression = expression;
        this.async = true;
    }
}
exports.AsyncArrowFunctionExpression = AsyncArrowFunctionExpression;
class AsyncFunctionDeclaration {
    constructor(id, params, body) {
        this.type = syntax_1.Syntax.FunctionDeclaration;
        this.id = id;
        this.params = params;
        this.body = body;
        this.generator = false;
        this.expression = false;
        this.async = true;
    }
}
exports.AsyncFunctionDeclaration = AsyncFunctionDeclaration;
class AsyncFunctionExpression {
    constructor(id, params, body) {
        this.type = syntax_1.Syntax.FunctionExpression;
        this.id = id;
        this.params = params;
        this.body = body;
        this.generator = false;
        this.expression = false;
        this.async = true;
    }
}
exports.AsyncFunctionExpression = AsyncFunctionExpression;
class AwaitExpression {
    constructor(argument) {
        this.type = syntax_1.Syntax.AwaitExpression;
        this.argument = argument;
    }
}
exports.AwaitExpression = AwaitExpression;
class BinaryExpression {
    constructor(operator, left, right) {
        const logical = (operator === '||' || operator === '&&');
        this.type = logical ? syntax_1.Syntax.LogicalExpression : syntax_1.Syntax.BinaryExpression;
        this.operator = operator;
        this.left = left;
        this.right = right;
    }
}
exports.BinaryExpression = BinaryExpression;
class BlockStatement {
    constructor(body) {
        this.type = syntax_1.Syntax.BlockStatement;
        this.body = body;
    }
}
exports.BlockStatement = BlockStatement;
class BreakStatement {
    constructor(label) {
        this.type = syntax_1.Syntax.BreakStatement;
        this.label = label;
    }
}
exports.BreakStatement = BreakStatement;
class CallExpression {
    constructor(callee, args) {
        this.type = syntax_1.Syntax.CallExpression;
        this.callee = callee;
        this.arguments = args;
    }
}
exports.CallExpression = CallExpression;
class CatchClause {
    constructor(param, body) {
        this.type = syntax_1.Syntax.CatchClause;
        this.param = param;
        this.body = body;
    }
}
exports.CatchClause = CatchClause;
class ClassBody {
    constructor(body) {
        this.type = syntax_1.Syntax.ClassBody;
        this.body = body;
    }
}
exports.ClassBody = ClassBody;
class ClassDeclaration {
    constructor(id, superClass, body) {
        this.type = syntax_1.Syntax.ClassDeclaration;
        this.id = id;
        this.superClass = superClass;
        this.body = body;
    }
}
exports.ClassDeclaration = ClassDeclaration;
class ClassExpression {
    constructor(id, superClass, body) {
        this.type = syntax_1.Syntax.ClassExpression;
        this.id = id;
        this.superClass = superClass;
        this.body = body;
    }
}
exports.ClassExpression = ClassExpression;
class ComputedMemberExpression {
    constructor(object, property) {
        this.type = syntax_1.Syntax.MemberExpression;
        this.computed = true;
        this.object = object;
        this.property = property;
    }
}
exports.ComputedMemberExpression = ComputedMemberExpression;
class ConditionalExpression {
    constructor(test, consequent, alternate) {
        this.type = syntax_1.Syntax.ConditionalExpression;
        this.test = test;
        this.consequent = consequent;
        this.alternate = alternate;
    }
}
exports.ConditionalExpression = ConditionalExpression;
class ContinueStatement {
    constructor(label) {
        this.type = syntax_1.Syntax.ContinueStatement;
        this.label = label;
    }
}
exports.ContinueStatement = ContinueStatement;
class DebuggerStatement {
    constructor() {
        this.type = syntax_1.Syntax.DebuggerStatement;
    }
}
exports.DebuggerStatement = DebuggerStatement;
class Directive {
    constructor(expression, directive) {
        this.type = syntax_1.Syntax.ExpressionStatement;
        this.expression = expression;
        this.directive = directive;
    }
}
exports.Directive = Directive;
class DoWhileStatement {
    constructor(body, test) {
        this.type = syntax_1.Syntax.DoWhileStatement;
        this.body = body;
        this.test = test;
    }
}
exports.DoWhileStatement = DoWhileStatement;
class EmptyStatement {
    constructor() {
        this.type = syntax_1.Syntax.EmptyStatement;
    }
}
exports.EmptyStatement = EmptyStatement;
class ExportAllDeclaration {
    constructor(source) {
        this.type = syntax_1.Syntax.ExportAllDeclaration;
        this.source = source;
    }
}
exports.ExportAllDeclaration = ExportAllDeclaration;
class ExportDefaultDeclaration {
    constructor(declaration) {
        this.type = syntax_1.Syntax.ExportDefaultDeclaration;
        this.declaration = declaration;
    }
}
exports.ExportDefaultDeclaration = ExportDefaultDeclaration;
class ExportNamedDeclaration {
    constructor(declaration, specifiers, source) {
        this.type = syntax_1.Syntax.ExportNamedDeclaration;
        this.declaration = declaration;
        this.specifiers = specifiers;
        this.source = source;
    }
}
exports.ExportNamedDeclaration = ExportNamedDeclaration;
class ExportSpecifier {
    constructor(local, exported) {
        this.type = syntax_1.Syntax.ExportSpecifier;
        this.exported = exported;
        this.local = local;
    }
}
exports.ExportSpecifier = ExportSpecifier;
class ExpressionStatement {
    constructor(expression) {
        this.type = syntax_1.Syntax.ExpressionStatement;
        this.expression = expression;
    }
}
exports.ExpressionStatement = ExpressionStatement;
class ForInStatement {
    constructor(left, right, body) {
        this.type = syntax_1.Syntax.ForInStatement;
        this.left = left;
        this.right = right;
        this.body = body;
        this.each = false;
    }
}
exports.ForInStatement = ForInStatement;
class ForOfStatement {
    constructor(left, right, body) {
        this.type = syntax_1.Syntax.ForOfStatement;
        this.left = left;
        this.right = right;
        this.body = body;
    }
}
exports.ForOfStatement = ForOfStatement;
class ForStatement {
    constructor(init, test, update, body) {
        this.type = syntax_1.Syntax.ForStatement;
        this.init = init;
        this.test = test;
        this.update = update;
        this.body = body;
    }
}
exports.ForStatement = ForStatement;
class FunctionDeclaration {
    constructor(id, params, body, generator) {
        this.type = syntax_1.Syntax.FunctionDeclaration;
        this.id = id;
        this.params = params;
        this.body = body;
        this.generator = generator;
        this.expression = false;
        this.async = false;
    }
}
exports.FunctionDeclaration = FunctionDeclaration;
class FunctionExpression {
    constructor(id, params, body, generator) {
        this.type = syntax_1.Syntax.FunctionExpression;
        this.id = id;
        this.params = params;
        this.body = body;
        this.generator = generator;
        this.expression = false;
        this.async = false;
    }
}
exports.FunctionExpression = FunctionExpression;
class Identifier {
    constructor(name) {
        this.type = syntax_1.Syntax.Identifier;
        this.name = name;
    }
}
exports.Identifier = Identifier;
class IfStatement {
    constructor(test, consequent, alternate) {
        this.type = syntax_1.Syntax.IfStatement;
        this.test = test;
        this.consequent = consequent;
        this.alternate = alternate;
    }
}
exports.IfStatement = IfStatement;
class Import {
    constructor() {
        this.type = syntax_1.Syntax.Import;
    }
}
exports.Import = Import;
class ImportDeclaration {
    constructor(specifiers, source) {
        this.type = syntax_1.Syntax.ImportDeclaration;
        this.specifiers = specifiers;
        this.source = source;
    }
}
exports.ImportDeclaration = ImportDeclaration;
class ImportDefaultSpecifier {
    constructor(local) {
        this.type = syntax_1.Syntax.ImportDefaultSpecifier;
        this.local = local;
    }
}
exports.ImportDefaultSpecifier = ImportDefaultSpecifier;
class ImportNamespaceSpecifier {
    constructor(local) {
        this.type = syntax_1.Syntax.ImportNamespaceSpecifier;
        this.local = local;
    }
}
exports.ImportNamespaceSpecifier = ImportNamespaceSpecifier;
class ImportSpecifier {
    constructor(local, imported) {
        this.type = syntax_1.Syntax.ImportSpecifier;
        this.local = local;
        this.imported = imported;
    }
}
exports.ImportSpecifier = ImportSpecifier;
class LabeledStatement {
    constructor(label, body) {
        this.type = syntax_1.Syntax.LabeledStatement;
        this.label = label;
        this.body = body;
    }
}
exports.LabeledStatement = LabeledStatement;
class Literal {
    constructor(value, raw) {
        this.type = syntax_1.Syntax.Literal;
        this.value = value;
        this.raw = raw;
    }
}
exports.Literal = Literal;
class MetaProperty {
    constructor(meta, property) {
        this.type = syntax_1.Syntax.MetaProperty;
        this.meta = meta;
        this.property = property;
    }
}
exports.MetaProperty = MetaProperty;
class MethodDefinition {
    constructor(key, computed, value, kind, isStatic) {
        this.type = syntax_1.Syntax.MethodDefinition;
        this.key = key;
        this.computed = computed;
        this.value = value;
        this.kind = kind;
        this.static = isStatic;
    }
}
exports.MethodDefinition = MethodDefinition;
class Module {
    constructor(body) {
        this.type = syntax_1.Syntax.Program;
        this.body = body;
        this.sourceType = 'module';
    }
}
exports.Module = Module;
class NewExpression {
    constructor(callee, args) {
        this.type = syntax_1.Syntax.NewExpression;
        this.callee = callee;
        this.arguments = args;
    }
}
exports.NewExpression = NewExpression;
class ObjectExpression {
    constructor(properties) {
        this.type = syntax_1.Syntax.ObjectExpression;
        this.properties = properties;
    }
}
exports.ObjectExpression = ObjectExpression;
class ObjectPattern {
    constructor(properties) {
        this.type = syntax_1.Syntax.ObjectPattern;
        this.properties = properties;
    }
}
exports.ObjectPattern = ObjectPattern;
class Property {
    constructor(kind, key, computed, value, method, shorthand) {
        this.type = syntax_1.Syntax.Property;
        this.key = key;
        this.computed = computed;
        this.value = value;
        this.kind = kind;
        this.method = method;
        this.shorthand = shorthand;
    }
}
exports.Property = Property;
class RegexLiteral {
    constructor(value, raw, pattern, flags) {
        this.type = syntax_1.Syntax.Literal;
        this.value = value;
        this.raw = raw;
        this.regex = { pattern, flags };
    }
}
exports.RegexLiteral = RegexLiteral;
class RestElement {
    constructor(argument) {
        this.type = syntax_1.Syntax.RestElement;
        this.argument = argument;
    }
}
exports.RestElement = RestElement;
class ReturnStatement {
    constructor(argument) {
        this.type = syntax_1.Syntax.ReturnStatement;
        this.argument = argument;
    }
}
exports.ReturnStatement = ReturnStatement;
class Script {
    constructor(body) {
        this.type = syntax_1.Syntax.Program;
        this.body = body;
        this.sourceType = 'script';
    }
}
exports.Script = Script;
class SequenceExpression {
    constructor(expressions) {
        this.type = syntax_1.Syntax.SequenceExpression;
        this.expressions = expressions;
    }
}
exports.SequenceExpression = SequenceExpression;
class SpreadElement {
    constructor(argument) {
        this.type = syntax_1.Syntax.SpreadElement;
        this.argument = argument;
    }
}
exports.SpreadElement = SpreadElement;
class StaticMemberExpression {
    constructor(object, property) {
        this.type = syntax_1.Syntax.MemberExpression;
        this.computed = false;
        this.object = object;
        this.property = property;
    }
}
exports.StaticMemberExpression = StaticMemberExpression;
class Super {
    constructor() {
        this.type = syntax_1.Syntax.Super;
    }
}
exports.Super = Super;
class SwitchCase {
    constructor(test, consequent) {
        this.type = syntax_1.Syntax.SwitchCase;
        this.test = test;
        this.consequent = consequent;
    }
}
exports.SwitchCase = SwitchCase;
class SwitchStatement {
    constructor(discriminant, cases) {
        this.type = syntax_1.Syntax.SwitchStatement;
        this.discriminant = discriminant;
        this.cases = cases;
    }
}
exports.SwitchStatement = SwitchStatement;
class TaggedTemplateExpression {
    constructor(tag, quasi) {
        this.type = syntax_1.Syntax.TaggedTemplateExpression;
        this.tag = tag;
        this.quasi = quasi;
    }
}
exports.TaggedTemplateExpression = TaggedTemplateExpression;
class TemplateElement {
    constructor(value, tail) {
        this.type = syntax_1.Syntax.TemplateElement;
        this.value = value;
        this.tail = tail;
    }
}
exports.TemplateElement = TemplateElement;
class TemplateLiteral {
    constructor(quasis, expressions) {
        this.type = syntax_1.Syntax.TemplateLiteral;
        this.quasis = quasis;
        this.expressions = expressions;
    }
}
exports.TemplateLiteral = TemplateLiteral;
class ThisExpression {
    constructor() {
        this.type = syntax_1.Syntax.ThisExpression;
    }
}
exports.ThisExpression = ThisExpression;
class ThrowStatement {
    constructor(argument) {
        this.type = syntax_1.Syntax.ThrowStatement;
        this.argument = argument;
    }
}
exports.ThrowStatement = ThrowStatement;
class TryStatement {
    constructor(block, handler, finalizer) {
        this.type = syntax_1.Syntax.TryStatement;
        this.block = block;
        this.handler = handler;
        this.finalizer = finalizer;
    }
}
exports.TryStatement = TryStatement;
class UnaryExpression {
    constructor(operator, argument) {
        this.type = syntax_1.Syntax.UnaryExpression;
        this.operator = operator;
        this.argument = argument;
        this.prefix = true;
    }
}
exports.UnaryExpression = UnaryExpression;
class UpdateExpression {
    constructor(operator, argument, prefix) {
        this.type = syntax_1.Syntax.UpdateExpression;
        this.operator = operator;
        this.argument = argument;
        this.prefix = prefix;
    }
}
exports.UpdateExpression = UpdateExpression;
class VariableDeclaration {
    constructor(declarations, kind) {
        this.type = syntax_1.Syntax.VariableDeclaration;
        this.declarations = declarations;
        this.kind = kind;
    }
}
exports.VariableDeclaration = VariableDeclaration;
class VariableDeclarator {
    constructor(id, init) {
        this.type = syntax_1.Syntax.VariableDeclarator;
        this.id = id;
        this.init = init;
    }
}
exports.VariableDeclarator = VariableDeclarator;
class WhileStatement {
    constructor(test, body) {
        this.type = syntax_1.Syntax.WhileStatement;
        this.test = test;
        this.body = body;
    }
}
exports.WhileStatement = WhileStatement;
class WithStatement {
    constructor(object, body) {
        this.type = syntax_1.Syntax.WithStatement;
        this.object = object;
        this.body = body;
    }
}
exports.WithStatement = WithStatement;
class YieldExpression {
    constructor(argument, delegate) {
        this.type = syntax_1.Syntax.YieldExpression;
        this.argument = argument;
        this.delegate = delegate;
    }
}
exports.YieldExpression = YieldExpression;

File: ts/scala/src/nodes.scala
/*
ScalaFromJS: Dev
nodes.ts
*/

package com.github.opengrabeso.esprima
/* import { Syntax } from './syntax' */
type ArgumentListElement = Expression | SpreadElement
type ArrayExpressionElement = Expression | SpreadElement
type ArrayPatternElement = AssignmentPattern | Identifier | ArrayPattern | ObjectPattern | RestElement
type BindingPattern = ArrayPattern | ObjectPattern
type BindingIdentifier = Identifier
type ChainElement = CallExpression | ComputedMemberExpression | StaticMemberExpression
type Declaration = Any
type ExportableDefaultDeclaration = Any
type ExportableNamedDeclaration = AsyncFunctionDeclaration | ClassDeclaration | FunctionDeclaration | VariableDeclaration
type ExportDeclaration = ExportAllDeclaration | ExportDefaultDeclaration | ExportNamedDeclaration
type Expression = Any
type FunctionParameter = AssignmentPattern | Identifier | ArrayPattern | ObjectPattern
type ImportDeclarationSpecifier = ImportDefaultSpecifier | ImportNamespaceSpecifier | ImportSpecifier
type ObjectExpressionProperty = Property | SpreadElement
type ObjectPatternProperty = Property | RestElement
type Statement = Any
type PropertyKey = Identifier | Literal
type PropertyValue = Any
type StatementListItem = Declaration | Statement
class ArrayExpression(var elements: Array[Expression | SpreadElement]) {
  var `type`: String = Syntax.ArrayExpression
}

class ArrayPattern(var elements: Array[AssignmentPattern | BindingIdentifier | BindingPattern | RestElement]) {
  var `type`: String = Syntax.ArrayPattern
}

class ArrowFunctionExpression(var params: Array[AssignmentPattern | Identifier | ArrayPattern | ObjectPattern], var body: BlockStatement | Expression, var expression: Boolean) {
  var `type`: String = Syntax.ArrowFunctionExpression
  var id: Identifier = null
  var generator: Boolean = false
  var async: Boolean = false
}

class AssignmentExpression(var operator: String, var left: Expression, var right: Expression) {
  var `type`: String = Syntax.AssignmentExpression
}

class AssignmentPattern(var left: Identifier | ArrayPattern | ObjectPattern, var right: Expression) {
  var `type`: String = Syntax.AssignmentPattern
}

class AsyncArrowFunctionExpression(var params: Array[AssignmentPattern | Identifier | ArrayPattern | ObjectPattern], var body: BlockStatement | Expression, var expression: Boolean) {
  var `type`: String = Syntax.ArrowFunctionExpression
  var id: Identifier = null
  var generator: Boolean = false
  var async: Boolean = true
}

class AsyncFunctionDeclaration(var id: Identifier, var params: Array[AssignmentPattern | Identifier | ArrayPattern | ObjectPattern], var body: BlockStatement, var generator: Boolean) {
  var `type`: String = Syntax.FunctionDeclaration
  var expression: Boolean = false
  var async: Boolean = true
}

class AsyncFunctionExpression(var id: Identifier, var params: Array[AssignmentPattern | Identifier | ArrayPattern | ObjectPattern], var body: BlockStatement, var generator: Boolean) {
  var `type`: String = Syntax.FunctionExpression
  var expression: Boolean = false
  var async: Boolean = true
}

class AwaitExpression(var argument: Expression) {
  var `type`: String = Syntax.AwaitExpression
}

class BinaryExpression(operator_par: String, left_par: Expression, right_par: Expression) {
  var `type`: String = _
  var operator: String = _
  var left: Expression = _
  var right: Expression = _
  this.constructor(operator_par, left_par, right_par)

  def constructor(operator: String, left: Expression, right: Expression) = {
    val logical = operator == "||" || operator == "&&" || operator == "??"
    this.`type` = if (logical) Syntax.LogicalExpression else Syntax.BinaryExpression
    this.operator = operator
    this.left = left
    this.right = right
  }
  
}

class BlockStatement(var body: Array[Statement]) {
  var `type`: String = Syntax.BlockStatement
}

class BreakStatement(var label: Identifier) {
  var `type`: String = Syntax.BreakStatement
}

class CallExpression(var callee: Expression | Import, var arguments: Array[Expression | SpreadElement], var optional: Boolean) {
  var `type`: String = Syntax.CallExpression
}

class CatchClause(var param: Identifier | ArrayPattern | ObjectPattern, var body: BlockStatement) {
  var `type`: String = Syntax.CatchClause
}

class ChainExpression(var expression: CallExpression | ComputedMemberExpression | StaticMemberExpression) {
  var `type`: String = Syntax.ChainExpression
}

class ClassBody(var body: Array[Property]) {
  var `type`: String = Syntax.ClassBody
}

class ClassDeclaration(var id: Identifier, var superClass: Identifier, var body: ClassBody) {
  var `type`: String = Syntax.ClassDeclaration
}

class ClassExpression(var id: Identifier, var superClass: Identifier, var body: ClassBody) {
  var `type`: String = Syntax.ClassExpression
}

class ComputedMemberExpression(var `object`: Expression, var property: Expression, var optional: Boolean) {
  var `type`: String = Syntax.MemberExpression
  var computed: Boolean = true
}

class ConditionalExpression(var test: Expression, var consequent: Expression, var alternate: Expression) {
  var `type`: String = Syntax.ConditionalExpression
}

class ContinueStatement(var label: Identifier) {
  var `type`: String = Syntax.ContinueStatement
}

class DebuggerStatement {
  var `type`: String = Syntax.DebuggerStatement
}

class Directive(var expression: Expression, var directive: String) {
  var `type`: String = Syntax.ExpressionStatement
}

class DoWhileStatement(var body: Statement, var test: Expression) {
  var `type`: String = Syntax.DoWhileStatement
}

class EmptyStatement {
  var `type`: String = Syntax.EmptyStatement
}

class ExportAllDeclaration(var source: Literal) {
  var `type`: String = Syntax.ExportAllDeclaration
}

class ExportDefaultDeclaration(var declaration: ExportableDefaultDeclaration) {
  var `type`: String = Syntax.ExportDefaultDeclaration
}

class ExportNamedDeclaration(var declaration: AsyncFunctionDeclaration | ClassDeclaration | FunctionDeclaration | VariableDeclaration, var specifiers: Array[ExportSpecifier], var source: Literal) {
  var `type`: String = Syntax.ExportNamedDeclaration
}

class ExportSpecifier(var local: Identifier, var exported: Identifier) {
  var `type`: String = Syntax.ExportSpecifier
}

class ExpressionStatement(var expression: Expression) {
  var `type`: String = Syntax.ExpressionStatement
}

class ForInStatement(var left: Expression, var right: Expression, var body: Statement) {
  var `type`: String = Syntax.ForInStatement
  var each: Boolean = false
}

class ForOfStatement(var left: Expression, var right: Expression, var body: Statement, var await: Boolean) {
  var `type`: String = Syntax.ForOfStatement
}

class ForStatement(var init: Expression, var test: Expression, var update: Expression, var body: Statement) {
  var `type`: String = Syntax.ForStatement
}

class FunctionDeclaration(var id: Identifier, var params: Array[AssignmentPattern | Identifier | ArrayPattern | ObjectPattern], var body: BlockStatement, var generator: Boolean) {
  var `type`: String = Syntax.FunctionDeclaration
  var expression: Boolean = false
  var async: Boolean = false
}

class FunctionExpression(var id: Identifier, var params: Array[AssignmentPattern | Identifier | ArrayPattern | ObjectPattern], var body: BlockStatement, var generator: Boolean) {
  var `type`: String = Syntax.FunctionExpression
  var expression: Boolean = false
  var async: Boolean = false
}

class Identifier(var name: String) {
  var `type`: String = Syntax.Identifier
}

class IfStatement(var test: Expression, var consequent: Statement, var alternate: Statement) {
  var `type`: String = Syntax.IfStatement
}

class Import {
  var `type`: String = Syntax.Import
}

class ImportDeclaration(var specifiers: Array[ImportDefaultSpecifier | ImportNamespaceSpecifier | ImportSpecifier], var source: Literal) {
  var `type`: String = Syntax.ImportDeclaration
}

class ImportDefaultSpecifier(var local: Identifier) {
  var `type`: String = Syntax.ImportDefaultSpecifier
}

class ImportNamespaceSpecifier(var local: Identifier) {
  var `type`: String = Syntax.ImportNamespaceSpecifier
}

class ImportSpecifier(var local: Identifier, var imported: Identifier) {
  var `type`: String = Syntax.ImportSpecifier
}

class LabeledStatement(var label: Identifier, var body: Statement) {
  var `type`: String = Syntax.LabeledStatement
}

class Literal(var value: Boolean | Double | String, var raw: String) {
  var `type`: String = Syntax.Literal
}

class MetaProperty(var meta: Identifier, var property: Identifier) {
  var `type`: String = Syntax.MetaProperty
}

class MethodDefinition(var key: Expression, var computed: Boolean, var value: AsyncFunctionExpression | FunctionExpression, var kind: String, var static: Boolean) {
  var `type`: String = Syntax.MethodDefinition
}

class Module(var body: Array[Declaration | Statement]) {
  var `type`: String = Syntax.Program
  var sourceType: String = "module"
}

class NewExpression(var callee: Expression, var arguments: Array[Expression | SpreadElement]) {
  var `type`: String = Syntax.NewExpression
}

class ObjectExpression(var properties: Array[Property | SpreadElement]) {
  var `type`: String = Syntax.ObjectExpression
}

class ObjectPattern(var properties: Array[Property | RestElement]) {
  var `type`: String = Syntax.ObjectPattern
}

class Property(var kind: String, var key: Identifier | Literal, var computed: Boolean, var value: PropertyValue, var method: Boolean, var shorthand: Boolean) {
  var `type`: String = Syntax.Property
}

class RegexLiteral(var value: (String) => Any, var raw: String, pattern: String, flags: String) {
  var `type`: String = Syntax.Literal
  var regex = new /*RegexLiteral/regex*/ {
    var pattern = pattern
    var flags = flags
  }
}

class RestElement(var argument: Identifier | ArrayPattern | ObjectPattern) {
  var `type`: String = Syntax.RestElement
}

class ReturnStatement(var argument: Expression) {
  var `type`: String = Syntax.ReturnStatement
}

class Script(var body: Array[Declaration | Statement]) {
  var `type`: String = Syntax.Program
  var sourceType: String = "script"
}

class SequenceExpression(var expressions: Array[Expression]) {
  var `type`: String = Syntax.SequenceExpression
}

class SpreadElement(var argument: Expression) {
  var `type`: String = Syntax.SpreadElement
}

class StaticMemberExpression(var `object`: Expression, var property: Expression, var optional: Boolean) {
  var `type`: String = Syntax.MemberExpression
  var computed: Boolean = false
}

class Super {
  var `type`: String = Syntax.Super
}

class SwitchCase(var test: Expression, var consequent: Array[Statement]) {
  var `type`: String = Syntax.SwitchCase
}

class SwitchStatement(var discriminant: Expression, var cases: Array[SwitchCase]) {
  var `type`: String = Syntax.SwitchStatement
}

class TaggedTemplateExpression(var tag: Expression, var quasi: TemplateLiteral) {
  var `type`: String = Syntax.TaggedTemplateExpression
}

trait TemplateElementValue {
  var cooked: String = _
  var raw: String = _
}

class TemplateElement(var value: TemplateElementValue, var tail: Boolean) {
  var `type`: String = Syntax.TemplateElement
}

class TemplateLiteral(var quasis: Array[TemplateElement], var expressions: Array[Expression]) {
  var `type`: String = Syntax.TemplateLiteral
}

class ThisExpression {
  var `type`: String = Syntax.ThisExpression
}

class ThrowStatement(var argument: Expression) {
  var `type`: String = Syntax.ThrowStatement
}

class TryStatement(var block: BlockStatement, var handler: CatchClause, var finalizer: BlockStatement) {
  var `type`: String = Syntax.TryStatement
}

class UnaryExpression(var operator: String, var argument: Expression) {
  var `type`: String = Syntax.UnaryExpression
  var prefix: Boolean = true
}

class UpdateExpression(var operator: String, var argument: Expression, var prefix: Boolean) {
  var `type`: String = Syntax.UpdateExpression
}

class VariableDeclaration(var declarations: Array[VariableDeclarator], var kind: String) {
  var `type`: String = Syntax.VariableDeclaration
}

class VariableDeclarator(var id: Identifier | ArrayPattern | ObjectPattern, var init: Expression) {
  var `type`: String = Syntax.VariableDeclarator
}

class WhileStatement(var test: Expression, var body: Statement) {
  var `type`: String = Syntax.WhileStatement
}

class WithStatement(var `object`: Expression, var body: Statement) {
  var `type`: String = Syntax.WithStatement
}

class YieldExpression(var argument: Expression, var delegate: Boolean) {
  var `type`: String = Syntax.YieldExpression
}

